src/v/serde/rw/variant.h

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/vlog.h"
+#include "serde/rw/rw.h"
+
+#include <array>
+#include <type_traits>
+#include <utility>
+#include <variant>
+
+namespace serde {
+
+// A small wrapper around std::variant that is marked to be serializable.
+//
+// Special precation needs to be taken to mark a variant type as serializable
+// with respect to compatibility. `serde::variant` should be a drop in
+// replacement for `std::variant`, but allows for serde operations.
+//
+// # Variant Wire Compatibility:
+//
+// Variants are tricky w.r.t. compatibility, they are forward compatible if new
+// discrimiants are appended to the variant.
+//
+// Concretely that means that you can only read old versions of a variant if new
+// options are appended. Writing a new variant option should happen with a
+// feature barrier, so that an older reader will never read a newly written
+// variant.
+template<typename... Types>
+struct variant : public std::variant<Types...> {
+    using type = std::variant<Types...>;
+
+    constexpr variant() = default;
+    constexpr variant(const variant&) = default;
+    constexpr variant(variant&&) = default;
+
+    // Ensure that this is not implicitly convertable from std::variant
+    template<class T>
+    constexpr variant(T&& t) noexcept // NOLINT(*-explicit-*)
+    requires(!std::is_same_v<std::decay_t<T>, std::variant<Types...>>)
+      : type(std::forward<T>(t)){};
+    explicit constexpr variant(std::variant<Types...> v) noexcept
+      : type(std::move(v)){};
```

Comment:
nit:
If I understand correctly, this is to prevent code like this to compile
```
auto var=serde::variant<int, bool>{10};
```

I think this pair of constructors can be condensed to

```suggestion
    template <std::same_as<std::variant<Types...>> SrcVar>
    explicit constexpr variant(SrcVar v) noexcept 
      : type(std::move(v)) {}
```

Reply:
Ah seems I can't negate things without a `requires` clause. I don't think this can be simplified unless I'm missing something